src/main/java/com/google/cloud/genomics/dataflow/pipelines/AnnotateVariants.java

```diff
@@ -0,0 +1,307 @@
+ * {@code VariantSet}s and {@code AnnotationSet}s as input and emits
+ * {@code VariantAnnotation}s. This program is intended to serve as an example
+ * of how a variant annotation program can be structured to run in parallel,
+ * operating over Google Genomics input sources; the current output has limited
+ * biological significance.
+ *
+ * Currently only annotates SNPs and uses the following two methods:
+ * <ul>
+ * <li>Determines the effect of the provided variants on the provided
+ *   transcripts, if any. A result is only emitted for a subset of cases where
+ *   the variant appears to cause a coding change in the transcript.
+ * <li>Performs an exact match join on the provided existing variant
+ *   annotations, if any.
+ * </ul>
+ */
+public final class AnnotateVariants extends DoFn<Contig, KV<String, VariantAnnotation>> {
+  private static final Logger LOG = Logger.getLogger(AnnotateVariants.class.getName());
+  private static final int VARIANTS_PAGE_SIZE = 400;
+  private static final String VARIANT_FIELDS
+      = "nextPageToken,variants(id,referenceName,start,end,alternateBases,referenceBases)";
+
+  private final GenomicsFactory.OfflineAuth auth;
+  private final String varsetId, refsetId;
+  private final List<String> callSetIds, transcriptSetIds, variantAnnotationSetIds;
+
+  public AnnotateVariants(GenomicsFactory.OfflineAuth auth, String varsetId, String refsetId,
+      List<String> callSetIds, List<String> transcriptSetIds,
+      List<String> variantAnnotationSetIds) {
+    this.auth = auth;
+    this.varsetId = varsetId;
+    this.refsetId = refsetId;
+    this.callSetIds = callSetIds;
+    this.transcriptSetIds = transcriptSetIds;
+    this.variantAnnotationSetIds = variantAnnotationSetIds;
+  }
+
+  @Override
+  public void processElement(
+      DoFn<Contig, KV<String, VariantAnnotation>>.ProcessContext c) throws Exception {
+    Genomics genomics = auth.getGenomics(auth.getDefaultFactory());
+
+    Contig contig = c.element();
+    IntervalTree<Annotation> transcripts = retrieveTranscripts(genomics, contig);
+    ListMultimap<Range<Long>, Annotation> variantAnnotations =
+        retrieveVariantAnnotations(genomics, contig);
+    long baseOffset = transcripts.min().getStart();
+    String refBases = retrieveReferenceBases(genomics, refsetId,
+        contig.referenceName, baseOffset, transcripts.max().getEnd());
+
+    int varCount = 0;
+    Iterable<Variant> varIter = FluentIterable
+        .from(Paginator.Variants.create(genomics, ShardBoundary.STRICT).search(
+            c.element().getVariantsRequest(varsetId)
+              // TODO: Variants-only retrieval is not well support currently. For now
+              // we parameterize by CallSet for performance.
+              .setCallSetIds(callSetIds)
+              .setPageSize(VARIANTS_PAGE_SIZE),
+            VARIANT_FIELDS))
+        .filter(VariantUtils.IS_SNP);
+    for (Variant variant : varIter) {
```

Comment:
CH, is right but you still want to keep track of what you're annotating since metadata is still important if you combine datasets or are comparing them.  If you can cache them that will save you time later on.